File: src/shady/generator/json_apply.c
#include <json-c/json.h>
#include <assert.h>

#include "log.h"

void json_apply_object(json_object* target, json_object* src);

void json_apply_object(json_object* target, json_object* src) {
    assert(target && src);
    assert(json_object_get_type(target) == json_type_object);
    assert(json_object_get_type(src) == json_type_object);
    struct json_object_iterator end = json_object_iter_end(src);
    for (struct json_object_iterator i = json_object_iter_begin(src); !json_object_iter_equal(&end, &i); json_object_iter_next(&i)) {
        const char* name = json_object_iter_peek_name(&i);
        json_object* value = json_object_iter_peek_value(&i);

        json_object* existing = json_object_object_get(target, name);
        if (existing && json_object_get_type(existing) == json_type_object) {
            json_apply_object(existing, value);
        } else if (existing && json_object_get_type(existing) == json_type_array && json_object_get_type(value) == json_type_array && json_object_array_length(value) <= json_object_array_length(existing)) {
            for (size_t j = 0; j < json_object_array_length(value); j++)
                json_object_array_put_idx(existing, j, json_object_array_get_idx(value, j));
        } else {
            if (existing)
                warn_print("json-apply: overwriting key '%s'\n", name);
            json_object_object_add(target, name, value);
            json_object_get(value);
        }
    }
}

File: src/generator/json_apply.py
import copy

def json_apply_object(target, src):
    assert (type(target) is dict)
    assert (type(src) is dict)
    for name, value in src.items():
        existing = target.get(name)
        if existing is not None and type(existing) is dict:
            json_apply_object(existing, value)
        elif existing is not None and type(existing) is list and type(value) is list:
            assert existing != value
            for elem in value:
                existing.append(elem)
        else:
            if existing is not None:
                print(f"json-apply: overwriting key '{name}'")
            target[name] = copy.deepcopy(value)